qualtran/bloqs/block_encoding/product.py

```diff
@@ -0,0 +1,264 @@
+                res_after = Register("res_after", dtype=QAny(res_bits_left))
+                res_regs = []
+                if res_bits_used > 0:
+                    res_regs.append(res_before)
+                res_regs.append(res)
+                res_bits_used += res_bits
+                if res_bits_left > 0:
+                    res_regs.append(res_after)
+                res_part = Partition(self.resource_bitsize, tuple(res_regs))
+                res_part_soqs = bb.add_d(res_part, x=soqs["resource"])
+                u_soqs["resource"] = res_part_soqs["res"]
+
+            # connect the constituent bloq
+            u_out_soqs = bb.add_d(u, **u_soqs)
+            system = u_out_soqs["system"]
+
+            # un-partition the resource register
+            if res_bits > 0:
+                res_part_soqs["res"] = u_out_soqs["resource"]
+                soqs["resource"] = cast(
+                    Soquet, bb.add(evolve(res_part, partition=False), **res_part_soqs)
+                )
+
+            # un-partition the ancilla register
+            if self.ancilla_bitsize > 0:
+                flag_bits_soq = cast(NDArray, anc_part_soqs["flag_bits"])
+                if anc_bits > 0:
+                    anc_used_soq = cast(Soquet, u_out_soqs["ancilla"])
+                    if i == n - 1:
+                        anc_part_soqs["anc_used"] = anc_used_soq
+                    else:
+                        # set corresponding flag if ancillas are all zero
+                        ctrl, flag_bits_soq[i] = bb.add_t(
+                            XGate().controlled(CtrlSpec(qdtypes=(anc_used_soq.reg.dtype,), cvs=0)),
+                            ctrl=anc_used_soq,
+                            q=flag_bits_soq[i],
+                        )
+                        flag_bits_soq[i] = bb.add(XGate(), q=flag_bits_soq[i])
+                        anc_part_soqs["anc_used"] = cast(Soquet, ctrl)
+                anc_part_soqs["flag_bits"] = flag_bits_soq
+                soqs["ancilla"] = cast(
+                    Soquet, bb.add(evolve(anc_part, partition=False), **anc_part_soqs)
+                )
+
+        out = {"system": system}
+        if self.ancilla_bitsize > 0:
+            out["ancilla"] = soqs["ancilla"]
+        if self.resource_bitsize > 0:
+            out["resource"] = soqs["resource"]
+        return out
+
+
+@bloq_example
+def _product_block_encoding() -> Product:
+    from qualtran.bloqs.basic_gates import Hadamard, TGate
+    from qualtran.bloqs.block_encoding.unitary import Unitary
+
+    product_block_encoding = Product((Unitary(TGate()), Unitary(Hadamard())))
+    return product_block_encoding
+
+
+@bloq_example
+def _product_block_encoding_override() -> Product:
+    from qualtran.bloqs.basic_gates import Hadamard, TGate
+    from qualtran.bloqs.block_encoding.unitary import Unitary
+
+    u1 = evolve(Unitary(TGate()), alpha=0.5, ancilla_bitsize=2, resource_bitsize=1, epsilon=0.01)
```

Comment:
why are you using evolve instead of just providing the arguments to the `Unitary` constructor?